### PR TITLE
MFAC Refactor

### PR DESCRIPTION
**Pull request by KSGulin**
Updated MFAC to work in the new pruning modifer framework
Removed MFAC options and instead propagated the args it held individually through the MFAC classes

### CODE REVIEW

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py [hunk 1]
```diff
@@ -598,9 +598,6 @@ def __init__(
             log_types=log_types,
             global_sparsity=global_sparsity,
             allow_reintroduction=allow_reintroduction,
-            init_sparsity=self._init_sparsity_orig,
-            final_sparsity=self._final_sparsity,
-            inter_func=inter_func,
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py:603]: these should be left in to be propagated as kwargs through `BasePruningModifier`, necessary for multiple inheritance with other classes

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py [hunk 2]
```diff
@@ -0,0 +1,1469 @@
...
+    |       final_sparsity: 0.8
+    |       start_epoch: 0.0
+    |       end_epoch: 10.0
+    |       update_frequency: 1.0
+    |       params: ["re:.*weight"]
+    |       leave_enabled: True
+    |       inter_func: cubic
+    |       log_types: __ALL__
+    |       mask_type: unstructured
+    |       mfac_options:
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py:89]: should update if we are getting rid of mfac_options

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py [hunk 3]
```diff
@@ -0,0 +1,1469 @@
...
+        inter_func: str = "cubic",
+        log_types: Union[str, List[str]] = ALL_TOKEN,
+        global_sparsity: bool = False,
+        use_gradient_buffering: Optional[bool] = None,
+        num_grads: Union[Dict[float, int], int] = 64,
+        damp: float = 1e-5,
+        grads_device: Union[str, int] = "cpu",
+        fisher_block_size: Optional[int] = 2000,
+        num_pages: int = 1,  # break computation into pages when block size is None
+        available_gpus: List[str] = field(default_factory=list),
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py:159]: need to add serializable ModifierProps for all of these

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py [hunk 4]
```diff
@@ -0,0 +1,1469 @@
...
+    "FisherInverseFast",
+    "FisherInverseFastBlock",
+    "FisherInverseFastPageSwap",
+    "FisherInverseFastSmallBlocks",
+]
+
+_LOGGER = logging.getLogger(__name__)
+BYTES_IN_MIB = 1024 ** 2
+@PyTorchModifierYAML()
+class MFACPruningModifier(BaseGradualPruningModifier):
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py:66]: consider inheriting from `GMPruningModifier` since the main change is just the scorer, then we get the ModifierProps for free

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py [hunk 5]
```diff
@@ -0,0 +1,1469 @@
...
+            self._last_applied_sparsity
+        )
+        self._grad_buffer = torch.zeros(
+            (num_grads, total_nonzero),
+            device="cpu",
+        )
+        self._buffer_idx = 0
+
+
+'''
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py:583]: nit: use `"""`

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py [hunk 6]
```diff
@@ -0,0 +1,1469 @@
...
+    ]
+
+
+def _block_memory_size(block_size: int, element_size: int) -> int:
+    """
+    Calculate memory needed for H^-1 calculations of one block.
+    """
+    # B^2 * e_size - memory required for H^-1
+    # 4*B * e_size - memory required for additional comp vectors
+    return (block_size ** 2 + 4 * block_size) * element_size
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_mfac.py:1469]: nit: no newline at end of file

File: src/sparseml/pytorch/sparsification/pruning/scorer.py [hunk 7]
```diff
@@ -97,11 +98,19 @@ class PruningParamsGradScorer(PruningParamsScorer, ABC):
...
+    :param grad_sampler: optional instance of GradSampler used to collect gradients
+        before pruning.
     """
 
-    def __init__(self, params: List[Parameter]):
+    def __init__(
+            self, 
+            params: List[Parameter],
+            grad_sampler: Optional[GradSampler] = None,
+        ):
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/scorer.py:109]: nit: over indented

File: src/sparseml/pytorch/sparsification/pruning/scorer.py [hunk 8]
```diff
@@ -97,11 +98,19 @@ class PruningParamsGradScorer(PruningParamsScorer, ABC):
...
     :param params: list of model Parameters to track and score
+    :param grad_sampler: optional instance of GradSampler used to collect gradients
+        before pruning.
     """
 
-    def __init__(self, params: List[Parameter]):
+    def __init__(
+            self, 
+            params: List[Parameter],
+            grad_sampler: Optional[GradSampler] = None,
```
Comment (bfineran) [inline src/sparseml/pytorch/sparsification/pruning/scorer.py:108]: are we going to add any shared logic for this?

File: src/sparseml/pytorch/utils/sparsification.py [hunk 9]
```diff
@@ -30,7 +39,10 @@
 )
 
 
-__all__ = ["ModuleSparsificationInfo"]
+__all__ = [
+    "ModuleSparsificationInfo",
+    "GradSampler",
+    ]
```
Comment (bfineran) [inline src/sparseml/pytorch/utils/sparsification.py:45]: nit: over indented

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+        )
+        == small_blocks_diag
+    )
+    assert (
+        pytest.approx(
+            fast_mul_out,
+            torch.sqrt(torch.mean(torch.square(fast_mul_out))).item() * PRECISION
+        )
+        == small_blocks_mul_out
+    )
```
Comment (bfineran) [inline tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:211]: nit: no newline at end of file

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+
+import pytest
+import torch
+
+from sparseml.pytorch.sparsification.pruning import (
+    FisherInverseFast,
+    FisherInverseFastBlock,
+    FisherInverseFastSmallBlocks,
+)
+
```
Comment (bfineran) [inline tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:25]: can we use `pytest.mark.parameterize` to reuse test code here? if for some reason we can't, let's move some of the shared code to a helper function

**Review by bfineran — CHANGES_REQUESTED** [review]

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py (hunk 1 above)
Comment (KSGulin) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_base.py:603]: Should these then be added to `BasePruningModifier`? Because right now they're not caught by any subclass, so they're propagated through to BaseObject, which throws an error.

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+
+# precent-wise precision in terms of the mean of the resulting tensors
+PRECISION = 0.000001
+
+
+@pytest.mark.skipif(
```
Comment (dbogunowicz) [inline tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:31]: nit: not used anymore

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac.py [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+    os.getenv("NM_ML_SKIP_PYTORCH_TESTS", False),
+    reason="Skipping pytorch tests",
+)
+def test_fast_small_blocks():
```
Comment (dbogunowicz) [inline tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:35]: Duplicating Ben's comment, let's pack it all inside `pytest.mark.parameterize`

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac.py [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+    assert (
+        pytest.approx(
+            fast_mul_out,
+            torch.mean(fast_mul_out).item() * PRECISION,
+        )
+        == small_blocks_mul_out
+    )
+
+
+@pytest.mark.skipif(
```
Comment (dbogunowicz) [inline tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:78]: nit: deprecated

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac.py (hunk 13 above)
Comment (dbogunowicz) [reply tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:35]: (`small_blocks`, `large_blocks` etc this could all be possibly one function with different args from `pytest.mark.parameterize`)

**Review by dbogunowicz — DISMISSED** [review]

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac.py (hunk 14 above)
Comment (bfineran) [reply tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:78]: not deprecated, just not main path. it's OK to keep in, especially for longer running tests

Comment (dbogunowicz) [reply tests/sparseml/pytorch/sparsification/pruning/test_mfac.py:78]: ok, now I get what is the purpose of this decorator. sorry for the confusion.

File: src/sparseml/pytorch/sparsification/pruning/scorer.py (hunk 8 above)
Comment (KSGulin) [reply src/sparseml/pytorch/sparsification/pruning/scorer.py]: I removed the grad sampler from here since this pathway isn't used. It may be worthwhile to create a shared pathway down the line if more gradient-based pruners get added.

File: tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_mfac.py [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+    )
+    assert isinstance(yaml_modifier, MFACPruningModifier)
+    pruning_modifier_serialization_vals_test(
+        yaml_modifier, serialized_modifier, obj_modifier, exclude_mask=True
+    )
+    assert (
+        str(yaml_modifier.global_sparsity)
+        == str(serialized_modifier.global_sparsity)
+        == str(obj_modifier.global_sparsity)
+    )
```
Comment (bfineran) [inline tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_mfac.py:333]: need to add assertions for all MFAC params here

**Review by bfineran — CHANGES_REQUESTED** [review]

**Review by bfineran — COMMENTED** [review]
LGTM, however tests are taking too long on GHA

**Review by markurtz — APPROVED** [review]

File: tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_mfac.py [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+    end_epoch = 15.0
+    update_frequency = 1.0
+    inter_func = "cubic"
+    global_sparsity = False
+    num_grads = 64
+    damp = 0.000001
+    grads_device = "cpu"
+    fisher_block_size = 20
+    num_pages = 1
+    available_devices = ["cuda:0"]
```
Comment (markurtz) [inline tests/sparseml/pytorch/sparsification/pruning/test_modifier_pruning_mfac.py:291]: Causing CPU only test failure

**Review by markurtz — CHANGES_REQUESTED** [review]
@KSGulin tests are failing locally for me when running on a CPU only machine (my MacBook). I believe this is likely why github actions is not completing do to some improper handling on their side. Ideally please update these to work on CPU as well as GPU. Worst case let's mark these tests as GPU only if we cannot get them to work on CPU only

File: tests/sparseml/pytorch/sparsification/pruning/test_mfac_inverse.py [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+    [
+        FisherInverseFastSmallBlocks,
+        FisherInverseFastBlock,
+    ],
+)
+@pytest.mark.parametrize(
+    "devices",
+    [
+        None,
+        ["cuda:0"],
```
Comment (markurtz) [inline tests/sparseml/pytorch/sparsification/pruning/test_mfac_inverse.py:47]: causing CPU only test failure

Comment (KSGulin) [reply tests/sparseml/pytorch/sparsification/pruning/test_mfac_inverse.py]: That explains it, thanks for catching that. I updated it so that pruning_mfac tests are all cpu based and the mfac_inverse tests that use gpu are skipped if a device is not found. The run pathways for cpu and gpu based runs diverge a good bit, which means that testing on a local machine with a GPU will test all the pathways in MFAC, but the tests on GHA unfortunately will not. Wanted to note this so that everyone is aware.

**Review by markurtz — APPROVED** [review]

**Review by bfineran — APPROVED** [review]
